### PR TITLE
Cstrs self adaptive

### PR DESCRIPTION
**Pull request by darioizzo**
A complete re-implementation of what in pagmo legacy had the same name. Several modifications were made so that performances can no longer be compared.

In particular the new version uses half fevals to compute the same and corrects a few bugs in the old code as well as taking care of corner cases.

### CODE REVIEW

File: include/pagmo/detail/custom_comparisons.hpp [hunk 1]
```diff
@@ -85,6 +87,33 @@ inline bool equal_to_f(T a, T b)
...
 
+// equal_to_vf than compares vectors of floating point types considering nan==nan
+template <typename T>
+struct equal_to_vf {
+    bool operator()(const std::vector<T> &lhs, const std::vector<T> &rhs) const
+    {
+        if (lhs.size() != rhs.size()) {
+            return false;
+        } else {
+            return std::equal(lhs.begin(), lhs.end(), rhs.begin(), equal_to_f<T>);
```
Comment (bluescarni) [inline include/pagmo/detail/custom_comparisons.hpp]: ``#include <algorithm>``

File: include/pagmo/detail/custom_comparisons.hpp [hunk 2]
```diff
@@ -85,6 +87,33 @@ inline bool equal_to_f(T a, T b)
...
+        }
+    }
+};
+
+// hash_vf can be used to hash vectors of floating point types
+template <typename T>
+struct hash_vf {
+    size_t operator()(std::vector<T> const &in) const
```
Comment (bluescarni) [inline include/pagmo/detail/custom_comparisons.hpp]: ``#include <cstddef>`` and use ``std::size_t``.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+ * depends on a given population, a method to update the penalties to a new population is provided.
+ *
+ * @see Farmani R., & Wright, J. A. (2003). Self-adaptive fitness formulation for constrained optimization.
+ * Evolutionary Computation, IEEE Transactions on, 7(5), 445-455 for the paper introducing the method.
+ *
+ */
+struct penalized_udp {
+public:
+    /// Unused default constructor to please the is_udp type trait
+    penalized_udp(){};
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: So this is never actually used in any possible case? If that is the case, it would probably be good to have an ``assert(false)`` in it.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 4]
```diff
@@ -0,0 +1,783 @@
...
+for more details.
+
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_ALGORITHMS_CSTRS_SELF_ADAPTIVE_HPP
+#define PAGMO_ALGORITHMS_CSTRS_SELF_ADAPTIVE_HPP
+
+#include <iomanip>
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: ``#include <cmath>``
``#include <cassert>``

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+#include <unordered_map>
+
+#include "../algorithm.hpp"
+#include "../algorithms/de.hpp"
+#include "../detail/custom_comparisons.hpp"
+#include "../exceptions.hpp"
+#include "../io.hpp"
+#include "../population.hpp"
+#include "../rng.hpp"
+#include "../utils/generic.hpp"
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: ``#include "../types.hpp"``

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+        }
+        return f;
+    }
+
+    // Call to this method updates all the members that are used to penalize the objective function
+    // As the penalization algorithm depends heavily on the ref population this method takes care of
+    // updating the necessary information. It also builds the hash map used to avoid unecessary fitness
+    // evaluations. We exclude this method from the test as all of its corner cases are difficult to trigger
+    // and test for correctness
+    // LCOV_EXCL_START
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: I feel like this is cheating. Normally we do this only for 3rd party code, which IMO is fine. But for our code, if it's really so difficult to cover all the test, at least we should pay the "iron price". Let's not hide the dust under the carpet :)

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+                }
+            }
+            // We call the evolution on the unconstrained population (here is where fevals will increase)
+            new_pop = static_cast<const algorithm *>(this)->evolve(new_pop);
+            penalized_udp_ptr = new_pop.get_problem().extract<detail::penalized_udp>();
+            // We update the original pop avoiding fevals thanks to the cache
+            for (decltype(pop.size()) i = 0u; i < pop.size(); ++i) {
+                auto x = new_pop.get_x()[i];
+                auto it_f = penalized_udp_ptr->m_fitness_map.find(x);
+                if (it_f != penalized_udp_ptr->m_fitness_map.end()) { // cash hit
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: ``cash -> cache`` I suppose

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 8]
```diff
@@ -0,0 +1,783 @@
...
+                    pop.set_xf(i, x, it_f->second);
+                } else { // we have to compute the fitness (this will increase the feval counter in the ref pop problem,
+                         // but should never happen hence we do not test it)
+                    pop.set_x(i, x); // LCOV_EXCL_LINE
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: Not sure I understand. If this branch never happens, can't we just eliminate it?

File: include/pagmo/algorithms/mbh.hpp [hunk 9]
```diff
@@ -366,7 +366,7 @@ class mbh : public algorithm
      * <tt>std::vector</tt> is a mbh::log_line_type containing: \p Fevals, \p Best, \p Violated, <tt>Viol. Norm</tt> and
      * \p Trial as described in mbh::set_verbosity().
      *
-     * @return an <tt> std::vector </tt> of mbh::log_line_type containing the logged values Fevals,
+     * @return an <tt> std::vector </tt> of mbh::log_line_type containing the logged values Fevals, Best,
```
Comment (bluescarni) [inline include/pagmo/algorithms/mbh.hpp]: The extra spaces around ``std::vector`` should be removed, they look ugly in the sphinx output.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp (hunk 8 above)
Comment (darioizzo) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: 1) Are we 100% sure the cahce as its implemented now will never fail? (before we had some cases where it could have returned unccorrectly the absence of a dv)
2) Are we 100% sure that we are never calling this on chromosomes previously unseen? (I think so, but did not want to bother to form in my mind a formal guarantee)

Anyway, will remove.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp (hunk 6 above)
Comment (darioizzo) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: No its not difficult to trigger all corner cases (just very very boring). But it is difficult to test their correctness without basically rewriting the code of the method update in the test cpp.

But ok yes, Its cheating ... will remove

File: include/pagmo/detail/custom_comparisons.hpp (hunk 1 above)
Comment (darioizzo) [inline include/pagmo/detail/custom_comparisons.hpp]: done ....

**Comment by darioizzo** [comment]
I think we should merge this as it contains a few things (like the doctest abilitation) and then I can complete the tests on a separate one.

File: pygmo/docstrings.cpp [hunk 10]
```diff
@@ -1494,6 +1510,117 @@ Get the perturbation vector.
...
+   > > Evolve the population using the UDA and a penalized objective
+   > > Reinsert the best decision vector from the previous evolution
+
+
+:class:`pygmo.core.cstrs_self_adaptive` is a user-defined algorithm (UDA) that can be used to construct :class:`pygmo.core.algorithm` objects.
+
+**NOTE** Self-adaptive constraints handling implements an internal cache to avoid the re-evaluation of the fitness
+for decision vectors already evaluated. This makes the final counter of function evaluations somehow unpredictable.
+The number of function evaluation will be bounded to \p iters times the fevals made by one call to the inner UDA. The
+internal cache is reset at each iteration, but its size will grow unlimited during each callto
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: callto -> call to

File: pygmo/docstrings.cpp [hunk 11]
```diff
@@ -1494,6 +1510,117 @@ Get the perturbation vector.
...
+the inner UDA evolve method.
+
+**NOTE** Several modification were made to the original Faramani and Wright ideas to allow their approach to work on
+corner cases and with any UDAs. Most notably, a violation to the \f$j\f$-th  constraint is ignored if all
+the decision vectors in the population satisfy that particular constraint (i.e. if \f$c_{j_{max}} = 0\f$).
+
+**NOTE** The performances of pagmo::cstrs_self_adaptive are highly dependent on the particular inner UDA employed and
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: missing period.

File: pygmo/docstrings.cpp [hunk 12]
```diff
@@ -1494,6 +1510,117 @@ Get the perturbation vector.
...
+)";
+}
+
+std::string cstrs_self_adaptive_get_log_docstring()
+{
+    return R"(get_log()
+
+Returns a log containing relevant parameters recorded during the last call to ``evolve()``. The log frequency depends on the verbosity parameter
+(by default nothing is logged) which can be set calling :func:`~pygmo.core.algorithm.set_verbosity()` on a :class:`~pygmo.core.algorithm` constructed
+with an :class:`~pygmo.core.mbh`. A verbosity level of ``N > 0`` will log one line each ``N`` ``iters``.
```
Comment (bluescarni) [inline pygmo/docstrings.cpp]: I presume the reference to ``mbh`` is wrong here?

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 13]
```diff
@@ -0,0 +1,798 @@
...
+     * @return the verbosity level.
+     */
+    unsigned get_verbosity() const
+    {
+        return m_verbosity;
+    }
+
+    /// Get log.
+    /**
+     * A log containing relevant quantities monitoring the last call to mbh::evolve(). Each element of the returned
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: ``mbh`` -> self adaptive I imagine.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 14]
```diff
@@ -0,0 +1,798 @@
...
+     * <tt>std::vector</tt> is a cstrs_self_adaptive::log_line_type containing: \p Iter, \p Fevals, \p Best, \p
+     * Infeasibility, \p Violated, <tt>Viol. Norm</tt>, <tt>N. Feasible</tt> as described in
+     * cstrs_self_adaptive::set_verbosity().
+     *
+     * @return an <tt>std::vector</tt> of cstrs_self_adaptive::log_line_type containing the logged values Iters,
+     * Fevals, Best, Infeasibility, Violated and Viol. Norm and N. Feasible
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: Missing period.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 15]
```diff
@@ -0,0 +1,798 @@
...
+ * The number of function evaluation will be bounded to \p iters times the fevals made by one call to the inner UDA. The
+ * internal cache is reset at each iteration, but its size will grow unlimited during each callto
+ * the inner UDA evolve method.
+ *
+ * **NOTE** Several modification were made to the original Faramani and Wright ideas to allow their approach to work on
+ * corner cases and with any UDAs. Most notably, a violation to the \f$j\f$-th  constraint is ignored if all
+ * the decision vectors in the population satisfy that particular constraint (i.e. if \f$c_{j_{max}} = 0\f$).
+ *
+ * **NOTE** The performances of pagmo::cstrs_self_adaptive are highly dependent on the particular inner UDA employed and
+ * in particular to its parameters (generations / iterations)
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: Missing period.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 16]
```diff
@@ -0,0 +1,798 @@
...
+    template <typename T, ctor_enabler<T> = 0>
+    explicit cstrs_self_adaptive(unsigned iters, T &&a, unsigned seed = pagmo::random_device::next())
+        : m_algorithm(std::forward<T>(a)), m_iters(iters), m_e(seed), m_seed(seed), m_verbosity(0u), m_log()
+    {
+    }
+
+    /// Evolve method.
+    /**
+     * This method will call evolve on the inner UDA \p iters times updating the penalty to be applied to the
+     * objective after each call
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: MP

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 17]
```diff
@@ -0,0 +1,798 @@
...
+     *     211           4190       -7.68336    0.000341894              1       0.273829              0 i
+     *     221           4390       -7.89941     0.00031154              1       0.273829              0 i
+     *     231           4590       -8.38299    0.000168309              1       0.147935              0 i
+     *     241           4790       -8.38299    0.000181461              1       0.147935              0 i
+     *     251           4989       -8.71021    0.000191197              1       0.100357              0 i
+     *     261           5189       -8.71021    0.000165734              1       0.100357              0 i
+     *     271           5389       -10.7421              0              0              0              3
+     *     281           5585       -10.7421              0              0              0              3
+     *     291           5784       -11.4868              0              0              0              4
+
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: I think an ``*`` is missing at the beginning of the line here.

File: include/pagmo/algorithms/cstrs_self_adaptive.hpp [hunk 18]
```diff
@@ -0,0 +1,798 @@
...
+    }
+    /// Set the algorithm verbosity.
+    /**
+     * This method will sets the verbosity level of the screen output and of the
+     * log returned by get_log(). \p level can be:
+     * - 0: no verbosity,
+     * - >0: will print and log one line each  \p level call to the inner algorithm.
+     *
+     * Example (verbosity 10):
+     * @code
```
Comment (bluescarni) [inline include/pagmo/algorithms/cstrs_self_adaptive.hpp]: This should probably be ``@code{.unparsed}`` for consistency.
